packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
     super(1);
   }
 }
+
+export class MultiProcessMutex {
+  // Logic explanation: the fs.writeFile function, when used with the wx+ flag, performs an atomic operation to create a file.
+  // If multiple processes try to create the same file simultaneously, only one will succeed.
+  // This logic can be utilized to implement a mutex.
+  // For more info check the Nomic Notion page (internal link).
+  // Timeout logic explanation:
+  // When the function passed to the mutex is executed, it must not take longer than _defaultTimeoutInMs.
+  // If it exceeds this duration, the function is terminated, and the mutex is released.
```

Comment:
I'm not sure we should do this. If a download is taking a long time for a valid reason (say, many compilers are being downloaded in a slow network), then this will interrupt the process unnecessarily. The only reason to do this is to handle the case where a download hangs, but this results in the same scenario than a process crashing, which should be handled anyway.

Reply:
I'll remove this logic for now, as we agreed via private messages